### ISSUE TITLE
Higher Resolution?

### ISSUE THREAD

**Issue by Pancakeddd** [davidson16807/tectonics.js#19]
Hey, i think it would be cool there was a little options box for choosing a higher resolution for the map.I  want to use this with world building, but the resolution is too small to be able to make it look real enough to use. Thanks!

**Comment by davidson16807** [comment]
Hm, should be doable. Everything's setup to work with arbitrary grids/resolutions as it is right now. Maybe this could be part of a larger feature for supplying options on startup: world radius, orbital parameters, solar luminosity, etc.

The architecture might even be flexible enough to switch grid resolution mid-simulation, but not a lot of payoff.

Current implementation uses a fibonacci grid, which looks pretty but requires cached data for quick startups. I'll probably revert to using icosahedrons if this feature gets implemented. 

Currently issue #17 is still outstanding and unless I get help I've got to solve that first.

**Comment by skepickle** [comment]
I've started working on this in my fork. I basically generated a Cached.js for myself with 50000 grid points. I didn't give it any controls though. I was thinking of making a querysting option that would generate/download new Cached.js files. OR maybe include the grid into a savefile. The problem is that it would make the save file huge. Maybe just save the # of points in save, and just live generate the grid/faces when loading or if a querystring option is present? I could work on that if you don't mind the lack of UI interface...

**Comment by davidson16807** [comment]
Awesome! I'll take a look at it. 

Important knowledge transfer:

 * Model can't currently support more than 65535 grid points. This is because I use Uint16Rasters to represent grid ids. 

 * I don't recommend adding the querystring for Cached.js. I don't think we should be building anything around Cached.js.  Cached.js is a hack that exists largely because THREEx.FibonacciSphereGeometry takes a looong time to load for a large number of grid points (>10000). I recommend switching the model over to IcosahedronGeometry, which will eliminate most of the need for Cached.js. We'll still need to generate VoronoiSphere, but that doesn't take nearly as long, maybe a second or so. Once we do that, we should be able to live generate grids of arbitrary size. 

 * We'll definitely need some way to represent grids in save files. Since we'll presumably be able to live generate grids, I recommend doing whatever you feel is easiest and not paying attention to save/load performance. 

 * I don't care about save file size. It was previously an issue when Estevio and I were trying to compress it down to a URL length, but it was too restrictive we abandoned that effort.  >1MB save files are completely reasonable.

 * I don't care about UI for now. Query string will do. Let's just keep taking steps in the right direction.

**Comment by skepickle** [comment]
I've started tackling this issue in earnest now. The changes needed to generate grid & Voronoi was easy (only done this locally). The grid generation is definitely slow. So I am going to try some experiments:

1. I will do as you suggested, and try IcosahedronGeometry.
2. I will try and see if I can add a control for changing the seed live. This way if someone loads the page with an alternate grid size, they don't have to re-generate the grid while experimenting with different seeds.
3. I will experiment with saving the grid into save-files, which will allow people to not have to regenerate the grid everytime they want to load . This will *definitely* blow up the size of the file, so I am also going to experiment with json compression. I will try [jsoncompress](https://www.npmjs.com/package/jsoncompress) first. It's a Node.js module, but it supports being run in the browser.

**Comment by davidson16807** [comment]
That all sounds good. I don't think compression will be needed but if you want go give it a shot. 

For #2, *it is totally possible to change resolution dynamically* Here's how. Let's say you have two grids of different resolution, grid1 and grid2. You want to convert from grid1 to grid2:

`grid1_ids_of_grid2_pos = grid1.getNearestIds(grid2.pos)` gives you a Uint16Raster that contains the ids of grid1 cells that are closest to the positions of grid2 cells. 

`grid2_raster Float32Raster.get_ids(grid1_raster, grid1_ids_of_grid2_pos)` gives you a Float32Raster that rescales grid1_raster to work with grid2. 

It's basically [nearest neighbor interpolation](https://en.wikipedia.org/wiki/Nearest-neighbor_interpolation).

**Comment by skepickle** [comment]
Yeah, so I was basically thinking this: We are feeding in the Random seed using a query-string. Which means if someone is exploring various seeds to see how the worlds look... then they have to reload each time. And if we're generating new grids when we load the page (instead of using what's in Cached), then it's a lot of wasted time. If, instead, the person loads the page using the grid-resolution as a query-string, and then can set the seed value using a text-box... then we can basically just reset the world data when they change the seed, but keep the grid/Voronoi in-place. How does that sound? Each time the seed gets modified:
1) time gets reset to 0 My.
2) all the plate data (and crust? i dunno the details) would get regenerated using the new Random object instance (with the new seed)...

**Comment by davidson16807** [comment]
I agree there should be a way to play with seed. Go ahead and work on it. But I think we need to set up a plan for a more general UI solution for setting initial conditions and parameters, seed and grid type being one of them.

I'm thinking we have a "New" button just above "Save". When you press the "New" button the simulation pauses, a new world is generated, and a sidebar pops out. The sidebar lets users enter initial conditions like planet radius or water fraction. Changing some parameters like "seed" will cause the world to regenerate, whereas other stuff will cause the existing planet to change dynamically. 

There might also be a need to change parameters mid-simulation. For instance, it might make sense to change orbital parameters mid simulation and see how climate is affected, but I think this would be a part of a separate dialog.

Thoughts?

**Comment by davidson16807** [comment]
As for regenerating world, you should just need to reinitialize the `world` object and feed it to model, as in `model.world(world)`. You can see this done in index.html.

**Comment by skepickle** [comment]
SO, I've been experimenting with the LZMA library. I have locally replaced the Cached.js's content with *compressed* versions of the exact same content. The file size goes from 1.4MB to 412kB. Why do we care? Well, the decompression at startup is fairly quick. I'm thinking... maybe I load up the Cached.js with an array of grid-face and voronoi cache data for various resolutions ranging from 10000 (current default) up to 65k (max uint16). I could make it so that only one that matches (or closest to) the requested resolution is decompressed. I know that this is kind of a hack for now. What do you think? Is this technical debt, or does it provide a useful option to a power user?

**Comment by davidson16807** [comment]
I tried speeding up VoronoiSphere generation, but turns out it didn't do much. Time to generate VoronoiSphere down from 20s to 10s. Yep, proceed with caching.

**Comment by skepickle** [comment]
LZMA can uncompress a cached 65k fib and voronai in 300ms. So yes. ;) We can basically provide several "presets" instead of a complete free-reign on resolution. We could maybe provide *another* option to let the user actually say "I don't care how long it takes, I want exact N vertices. ¯\_(ツ)_/¯

**Comment by davidson16807** [comment]
>We could maybe provide another option to let the user actually say "I don't care how long it takes, I want exact N vertices.

Meh, I don't think the user would care enough for us to complicate the UI

**Comment by davidson16807** [comment]
Made another attempt to speed up VoronoiSphere generation, this time leveraging the fact that points will never exceed a given distance. 40k planet generation is now down to 5 seconds, and I think I could combine the strategies to bring it down even further maybe to 2 seconds, but still not so great.

**Comment by davidson16807** [comment]
OK, scratch that. Got it down to 600ms for 40k. Now we're talking.

**Comment by skepickle** [comment]
I'm not dead. Still working on this.

**Comment by davidson16807** [comment]
Haha, alright. I appreciate your help on this. Were the performance improvements any help? Is there an obstacle you need me to unblock?

**Comment by skepickle** [comment]
No obstacles yet, other than my day-job ;-) I believe your performance improvement really help out! I can basically focus on getting a nice dialog window... like "Edit" next to Save/Load. And I get to learn how that works! Woo!

**Comment by skepickle** [comment]
So, I've been looking at the new IcosagedronGeometry, and anything above a "detail" level of 6 ends up just stalling out. Do you think this is because it generates too many vertices? I ask, because when the detail level is 6, it's definitely less vertices than when I set the Fib-grid up to 65k points. Is the # of points exponential to detail level? If so, I might need to bump the vertex-# from Uint16 to Uint32. :-D

**Comment by davidson16807** [comment]
Yes, level 7 has 163k vertices. It's definitely the Uint16 barrier.

    > foo = new THREE.IcosahedronGeometry(1, 7).vertices.length
    163842

We don't have any use for Uint16 besides storing grid ids, so a dumb find/replace that replaces "Uint16" with "Uint32" will let you break past the 65k barrier.

**Comment by skepickle** [comment]
That escalated quickly. :-D

**Comment by davidson16807** [comment]
Found a complication to working with higher resolutions:

Tectonics uses binary morphology functions to generate plate boundaries. These functions accept a `radius` parameter, where `radius` is measured in grid cells, not km or radians. In other words, if you increase the resolution, you'll get stranger looking plate boundaries.

Compare this (10k):

![bar](https://user-images.githubusercontent.com/5824741/36918055-2e4cf136-1e1e-11e8-8688-d0a83098573a.png)

with this (40k):

![foo](https://user-images.githubusercontent.com/5824741/36918032-26225104-1e1e-11e8-9202-c6d84935766e.png)

Two possible solutions:

 * completely overhaul plate boundary generation, drop use of image segmentation algorithm in favor of a gradient damage model (which I'm still not sure I fully understand), or...

 * adjust calls to binary morphology functions so that `radius` is defined in radians.

latter is the easiest to implement

**Comment by skepickle** [comment]
I understand your point (_but only marginally_.)  I'm not a plate tectonics expert, so the 40k image doesn't look "stranger" to me, but I probably don't know what to look for. That being said, I think I'm going to build in a non-default option to use Fibonacci grid generation. It is *much* slower, but gives linear scaling of the number of vertices, versus the Icosehedron which grows exponentially.

**Comment by davidson16807** [comment]
Yeah, it's not a deal breaker, it just violates an assumption I previously made where model behavior was invariant to grid resolution.

With fibonacci grids, you'll still have problems past 65k vertices (unless that was fixed a while back, can't remember), but if there's lots of performance problems for level 7 icosahedrons then I guess that's fine.

**Comment by skepickle** [comment]
I believe leaving icosahedron as the default makes sense. The fib grid should only be used if someone knows that option even exists...

**Comment by NJSwetye** [comment]
This is just brilliant. Been mucking around for a few days here after just lurking a few years back. 

Do I understand correctly, that 

`new` THREE.IcosahedronGeometry(1, 6), `

is set in Lithosphere.js 

I don't seem to be able to create much higher "resolution", even by changing the original value of 1,3 up to even 1,60 so I must be missing something.

**Comment by davidson16807** [comment]
Try setting this line in index.html:

`var geometry = new THREE.IcosahedronGeometry(1, resolution);`

I appreciate your interest in the source code. I think we've amassed more issues than one developer can take on right now and we could use some help, even if that's just another set of eyes on the code.

**Comment by NJSwetye** [comment]
I wish I could be of more help than I probably can. I've done some extensive modding of games, but really diving into the code is beyond my expertise. Mucking with values and testing, that I can do.

**Comment by davidson16807** [comment]
> Mucking with values and testing, that I can do.

You're certainly welcome to help, however you can! You seem to be adept with modding, would you like to poke around with the rasters framework?  There's a [devtool](http://davidson16807.github.io/tectonics.js/tools/devtool.html) you can use to mess around with scalar/vector rasters, and a list functions you can use [here](https://github.com/davidson16807/tectonics.js/blob/master/postcompiled/Rasters.js). Perhaps we can compile some documentation for an API?

**Comment by davidson16807** [comment]
Status update: I did some work recently on the user interface that addresses a lot of points made in this issue. If anyone's around still they can check it out here: https://htmlpreview.github.io/?https://github.com/davidson16807/tectonics.js/blob/master/index.html

Btw, either of you get the link to the discord?

**Comment by NJSwetye** [comment]
awesome - I'll check it out. You can send that discord link of you like

**Comment by davidson16807** [comment]
sure, try this one: https://discord.gg/gQ7Jhj2
